torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp

```diff
@@ -596,6 +596,87 @@ class TORCH_API ProcessGroupNCCL : public Backend {
     std::string traceKeyEnd_;
   };
 
+  // Class that runs as a separate thread aside from watchdog
+  // thread because we need to check the heartbeat from watchdog thread
+  // so that when we get stuck in some NCCL/CUDA calls,
+  // we can dump the debugging information and abort the process.
+  class HeartbeatMonitor {
+   public:
+    HeartbeatMonitor(ProcessGroupNCCL* pg);
```

Comment:
Use cpp way. `intrusive_ptr` or `shared_ptr`. 
Also, to avoid cyclic dependency problem, this shared ptr can be a weak pointer:
> A weak pointer is a type of smart pointer that allows you to hold a non-owning reference to an object that is owned by a shared pointer.

Reply:
Actually that's why we don't use intrusive_ptr or shared_ptr here to avoid a cyclic dependency. And a raw pointer is indeed is ok here because the life cycle of this class is always within PGNCCL so we won't have illegal access?